cpp/src/reductions/scan.cu

```diff
@@ -2,74 +2,149 @@
-                                      inp, out, size);
         CUDA_CHECK_LAST();
         return GDF_SUCCESS;
     }
 
-    static
-    gdf_error inclusive_sum(void *&temp_storage, size_t &temp_storage_bytes,
-                            const T *inp, T *out, size_t size) {
-        cub::DeviceScan::InclusiveSum(temp_storage, temp_storage_bytes,
-                                      inp, out, size);
-        CUDA_CHECK_LAST();
-        return GDF_SUCCESS;
-    }
-};
-
-#define SCAN_IMPL(F, T)                                                       \
-gdf_error gdf_prefixsum_##F(gdf_column *inp, gdf_column *out, int inclusive) {\
-    GDF_REQUIRE( inp->size == out->size, GDF_COLUMN_SIZE_MISMATCH );          \
-    GDF_REQUIRE( inp->dtype == out->dtype, GDF_UNSUPPORTED_DTYPE );           \
-    GDF_REQUIRE( !inp->valid || !inp->null_count, GDF_VALIDITY_UNSUPPORTED ); \
-    GDF_REQUIRE( !out->valid || !out->null_count, GDF_VALIDITY_UNSUPPORTED ); \
-    return Scan<T>::call((const T*)inp->data, (T*)out->data, inp->size,       \
-                         inclusive);                                          \
-}
+    template <class T>
+    struct Scan {
+        static
+            gdf_error call(const gdf_column *input, gdf_column *output, bool inclusive,
+                cudaStream_t stream) {
+            using cub::DeviceScan;
+            auto scan_function = (inclusive ? inclusive_sum : exclusive_sum);
+            size_t size = input->size;
+            const T* d_input = static_cast<const T*>(input->data);
+            T* d_output = static_cast<T*>(output->data);
+
+            // Prepare temp storage
+            void *temp_storage = NULL;
+            size_t temp_storage_bytes = 0;
+            scan_function(temp_storage, temp_storage_bytes, d_input, d_output,
+                size, stream);
+            RMM_TRY(RMM_ALLOC(&temp_storage, temp_storage_bytes, stream));
+
+            bool const input_has_nulls{ input->valid != nullptr && input->null_count > 0 };
+            if (input_has_nulls) {
+                // allocate temporary column data
+                T* temp_input;
+                RMM_TRY(RMM_ALLOC(&temp_input, size * sizeof(T), stream));
+
+                // copy bitmask
+                size_t valid_byte_length = gdf_get_num_chars_bitmask(size);
+                CUDA_TRY(cudaMemcpyAsync(output->valid, input->valid,
+                        valid_byte_length, cudaMemcpyDeviceToDevice, stream));
+                output->null_count = input->null_count;
 
+                // copy d_input data and replace with 0 if mask is not valid
+                copy_and_replace_nulls(
+                    static_cast<const T*>(input->data), input->valid,
+                    size, temp_input, static_cast<T>(0), stream);
 
-SCAN_IMPL(i8,  int8_t)
-SCAN_IMPL(i32, int32_t)
-SCAN_IMPL(i64, int64_t)
+                // Do scan
+                scan_function(temp_storage, temp_storage_bytes, temp_input,
```

Comment:
Need to check the returned error code.
 
Use `GDF_REQUIRE` to ensure it equals `GDF_SUCCESS`.